crates/pico/tests/memoization.rs

```diff
@@ -0,0 +1,93 @@
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        LazyLock, Mutex,
+    },
+};
+
+use calc::{ast::Program, error::Result, eval::eval, lexer::Lexer, parser::Parser};
+use pico::storage::DefaultStorage;
+use pico_core::{database::Database, source::SourceId, storage::Storage};
+use pico_macros::{memo, Db, Source};
+
+mod calc;
+
+static SUM_COUNTER: AtomicUsize = AtomicUsize::new(0);
+static EVAL_COUNTER: LazyLock<Mutex<HashMap<SourceId<Input>, usize>>> =
+    LazyLock::new(|| Mutex::new(HashMap::new()));
+
+#[test]
+fn memoization() {
+    let mut state = TestDatabase::default();
```

Comment:
test lgtm

can we also add an analogous test for when we update `left` and give it a new value, and assert that therefore `sum` was recalculated?

Reply:
Maybe add to the same test as a third step?

Reply:
yeah sgtm